### PR TITLE
ues cross entropy (rather than perplexity) to estimate reading times

### PR DESCRIPTION
**Pull request by mschrimpf**

### CODE REVIEW

**Review by hans — CHANGES_REQUESTED** [review]

File: brainscore_language/models/huggingface.py [hunk 1]
```diff
@@ -130,16 +130,16 @@ def digest_text(self, text: Union[str, List[str]]) -> Dict[str, DataAssembly]:
     def estimate_reading_times(self, base_output):
         """
         :param base_output: the neural network's output
-        :return: perplexity (calculated as cross entropy) as a proxy for reading times
+        :return: cross entropy as a proxy for reading times, following Smith & Levy 2013
+            (https://www.sciencedirect.com/science/article/pii/S0010027713000413)
         """
         import torch.nn.functional as F
         tokens = self.tokenized_inputs['input_ids'].squeeze()
         logits = base_output.logits.squeeze()
```
[the remark this answers is not in the document]
Comment (mschrimpf) [reply brainscore_language/models/huggingface.py:138]: Right, this was missed in https://github.com/brain-score/language/pull/35.

How do you compute the reading time of a single word in that case? I.e. I show the subject the word "Hello", it takes them 1 second to read, how do you compare to the model? Language generation from scratch?

File: brainscore_language/models/huggingface.py [hunk 2]
```diff
@@ -130,16 +130,16 @@ def digest_text(self, text: Union[str, List[str]]) -> Dict[str, DataAssembly]:
...
         # assume that reading time is additive,
         # i.e. reading time of multiple tokens is the sum of the perplexity of each individual token
         cross_entropy = F.cross_entropy(logits, tokens, reduction='sum')
```
[the remark this answers is not in the document]
Comment (mschrimpf) [reply brainscore_language/models/huggingface.py]: Yes, this would be up to the benchmark. The current implementation of Futrell2018 https://github.com/brain-score/language/blob/55a0793d8c1e00eab9d08ab5217009ba22026c03/brainscore_language/plugins/futrell2018/benchmark.py#L42-L43 goes word-by-word. Would you explicitly guard against this from the model side?

File: brainscore_language/models/huggingface.py [hunk 3]
```diff
@@ -130,16 +130,16 @@ def digest_text(self, text: Union[str, List[str]]) -> Dict[str, DataAssembly]:
     def estimate_reading_times(self, base_output):
```
[the remark this answers is not in the document]
Comment (mschrimpf) [reply brainscore_language/models/huggingface.py:130]: I thought the assumption was that the bits are an approximation of reading times, such that they are correlated even without additional mapping? Or do you usually still regress from bits to reading times?

[the remark this answers is not in the document]
Comment (hans) [reply brainscore_language/models/huggingface.py:130]: I just checked and a raw Pearson correlation between word-level surprisal and reading times (on two different types of reading time methods, eye-tracking and self-paced reading/SPRT) does indeed come out significant. So it's a good start.

But the norm is to learn a more complex linear regression, minimally involving
1. surprisal of the previous 2–4 words (depending on method; in SPRT, there are "spillover effects" so the effect of surprisal gets smeared over multiple word reading times)
2. control predictors: word length and frequency (of current and preceding 2–4 words)

Anyway, raw correlation is a good start, but if we're looking to get an upper-bound of a model's RT prediction performance, we'd want a regression model like this.

File: brainscore_language/models/huggingface.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (hans) [reply brainscore_language/models/huggingface.py:138]: Hi, I'm not sure I understand the question. Here's a slice of a reading time dataset:

image

We compute a column `surprisal` by feeding the same text to a model that was used for an RT corpus. We index into the model's softmax outputs to get its estimate of $p(w_i | w_{< i})$ for each ground-truth word $w_i$. (There's some messiness re: tokens vs. words here of course, but that's the basic setup.)

You can see an example of the same indexing job in the [`SyntaxGym implementation`](https://huggingface.co/spaces/cpllab/syntaxgym/blob/main/syntaxgym.py#L197):

```python
        # Compute surprisal per-batch and combine into a single surprisal tensor.
        n_sentences, n_timesteps = all_tokenized_sentences["input_ids"].shape
        surprisals = torch.zeros(n_sentences, n_timesteps - 1).float().to(device)
        for i, batch in enumerate(datasets) :
            batch = batch.to(device)
            with torch.no_grad():
                # logits are B * T * V
                b_logits = model(batch)["logits"]
                b_surprisals = -b_logits.log_softmax(dim=2) / np.log(2)

            # Get surprisals of ground-truth words.
            gt_idxs = batch[:, 1:]
            # Reindexed surprisals: B * (T - 1)
            b_surprisals_gt = torch.gather(b_surprisals[:, :-1, :], 2, gt_idxs.unsqueeze(2)).squeeze(2)

            surprisals[i * batch_size : (i + 1) * batch_size] = b_surprisals_gt
```

Note the reindexing on the time axis of `gt_idxs = batch[:, 1:]` and `b_surprisals[:, :-1, :]`.

File: brainscore_language/models/huggingface.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (hans) [reply brainscore_language/models/huggingface.py:130]: Maybe TMI, but here's what a full reading time regression model looks like. Just pulled one off the shelf at random. Note effects of the surprisal of both the current word and the previous word.

```
formula: psychometric ~ surprisal + prev_surp + prev2_surp + prev3_surp + 
    prev4_surp + freq + prev_freq + prev2_freq + prev3_freq + 
    prev4_freq + len + prev_len + prev2_len + prev3_len + prev4_len

Call:
lm(formula = formula, data = df)

Residuals:
   Min     1Q Median     3Q    Max 
-59.70 -21.46  -5.59  15.19 462.01 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)    
(Intercept)  3.036e+02  7.447e+00  40.764  < 2e-16 ***
surprisal    1.215e+00  2.307e-01   5.264 1.50e-07 ***
prev_surp    5.570e-01  2.285e-01   2.438  0.01483 *  
prev2_surp   1.840e-02  2.270e-01   0.081  0.93538    
prev3_surp  -8.355e-04  2.325e-01  -0.004  0.99713    
prev4_surp  -2.825e-01  2.210e-01  -1.278  0.20125    
freq        -3.312e-01  3.021e-01  -1.096  0.27309    
prev_freq   -2.586e-01  3.066e-01  -0.843  0.39916    
prev2_freq  -3.924e-01  3.102e-01  -1.265  0.20595    
prev3_freq   1.718e-01  3.014e-01   0.570  0.56875    
prev4_freq   1.577e-02  2.224e-01   0.071  0.94348    
len          1.643e+00  3.891e-01   4.222 2.49e-05 ***
prev_len     8.601e-01  3.841e-01   2.239  0.02523 *  
prev2_len    8.631e-01  3.862e-01   2.235  0.02551 *  
prev3_len    1.188e+00  3.851e-01   3.085  0.00205 ** 
prev4_len    8.307e-01  3.525e-01   2.357  0.01850 *  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 32.67 on 3090 degrees of freedom
Multiple R-squared:  0.0691,	Adjusted R-squared:  0.06458 
F-statistic: 15.29 on 15 and 3090 DF,  p-value: < 2.2e-16
```

[the remark this answers is not in the document]
Comment (mschrimpf) [reply brainscore_language/models/huggingface.py:130]: ok I would cast that into two options:
1. people can submit a model that already outputs reading times, e.g. by learning the regression from bits to RT on separate data
2. we introduce another benchmark (or change this one) to use a regression metric (from model-predicted RTs to human RTs)

Do you/Roger have a preference?

File: brainscore_language/models/huggingface.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mschrimpf) [reply brainscore_language/models/huggingface.py:138]: ah so the first couple words in the sequence are actually ignored here? ("**If you were to** journey to the North of ...")

File: brainscore_language/models/huggingface.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mschrimpf) [reply brainscore_language/models/huggingface.py]: looking at this again, the multiplication by `-1` in the context of reading times is weird since that will give us negative values. Do you typically not worry about reading times being negative? (I know you've been treating this as "proto reading times" that still need to be regressed to actual human reading times, but I would want something that is at least positive to avoid confusion for other users)

File: brainscore_language/models/huggingface.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (hans) [reply brainscore_language/models/huggingface.py:138]: That's not due to anything on the language modeling side -- that's because our RT regression model uses data about the surprisals of the past 4 words, which is of course undefined for the first 3 words in each corpus.

But it is true that we have to ditch the first token in the model's tokenized input, since by construction the model doesn't have a prediction for this item. But because models often have a beginning-of-sentence token, sometimes this doesn't matter.

File: brainscore_language/models/huggingface.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (hans) [reply brainscore_language/models/huggingface.py]: Oops, I forgot there's already a negative in the cross entropy formula. Yes, surprisal should be positive, so just `cross_entropy / np.log(2)` should be good.

File: brainscore_language/models/huggingface.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (hans) [reply brainscore_language/models/huggingface.py:137]: (For the record, we resolved in favor of \#2 in the meeting today.)

File: brainscore_language/models/huggingface.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mschrimpf) [reply brainscore_language/models/huggingface.py]: from zoom call: don't guard against it, but document clearly

File: brainscore_language/models/huggingface.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mschrimpf) [reply brainscore_language/models/huggingface.py:137]: from zoom call: build two benchmarks
1. linear predictivity metric
2. direct correlation
Model implementations then can choose to output any prediction they like

File: brainscore_language/models/huggingface.py [hunk 4]
```diff
@@ -99,59 +103,97 @@ def digest_text(self, text: Union[str, List[str]]) -> Dict[str, DataAssembly]:
...
+            representation_values,
+            coords={**stimuli_coords, **neuroid_coords},
+            dims=['presentation', 'neuroid'])
+        return representations
+
+    def estimate_reading_times(self, base_output: CausalLMOutput):
         """
         :param base_output: the neural network's output
-        :return: perplexity (calculated as cross entropy) as a proxy for reading times
+        :return: cross entropy as a proxy for reading times, following Smith & Levy 2013
```
Comment (hans) [inline brainscore_language/models/huggingface.py:152]: "surprisal (in bits) as a proxy for reading times"
(ambiguity over log-base can be an issue!)

File: brainscore_language/models/huggingface.py [hunk 5]
```diff
@@ -99,59 +103,97 @@ def digest_text(self, text: Union[str, List[str]]) -> Dict[str, DataAssembly]:
...
+        # get expectation (logits), shifted left by 1
+        predicted_logits = logits[-self.current_tokens['input_ids'].shape[-1] - 1: - 1, :].contiguous()
+        actual_tokens = self.current_tokens['input_ids'].squeeze(dim=0).contiguous()
+        first_token_addition = 0
+        if actual_tokens.shape[0] == predicted_logits.shape[0] + 1:  # multiple tokens for first model input
+            actual_tokens = actual_tokens[1:]  # we have no prior context to predict the 0th token
+            first_token_addition = FIRST_TOKEN_READING_TIME  # add the expected reading time for 0th token (see above)
+
+        # assume that reading time is additive, i.e. reading time of multiple tokens is
+        # the sum of the perplexity of each individual token.
```
Comment (hans) [inline brainscore_language/models/huggingface.py:174]: `s/perplexity/surprisal/`

File: brainscore_language/models/huggingface.py [hunk 6]
```diff
@@ -99,59 +103,97 @@ def digest_text(self, text: Union[str, List[str]]) -> Dict[str, DataAssembly]:
...
         pred_id = torch.argmax(logits, axis=2).squeeze()
+        # Note that this is currently only predicting the next *token* which might not always be entire words.
         last_model_token_inference = pred_id[-1].tolist()
         next_word = self.tokenizer.decode(last_model_token_inference)
-
+        # `next_word` often includes a space ` ` in front of the actual word. Since the task already tells us to output
+        # a word, we can strip the space.
+        # Note that the model might also predict a token completing the current word, e.g. `fox` -> `es` (`foxes`),
+        # this is not caught in the current implementation.
+        next_word = next_word.strip()
```
Comment (hans) [inline brainscore_language/models/huggingface.py:196]: Yeah, this is slightly hacky. The right way to do this would be to have the client accept the tokens as-is and then run the model's detokenizer on output (`PreTrainedTokenizer.decode`). But if the current setup is sufficient for our current use cases, then seems good!

File: tests/test_models/test_huggingface.py [hunk 7]
```diff
@@ -153,5 +154,3 @@ def test_one_text_two_targets(self):
 
     # TODO: add test with long text input, e.g. thousands of words,
     #  to see if we need batching, and to stress-test token alignment
```
Comment (hans) [inline tests/test_models/test_huggingface.py:156]: Other edge cases I'd want to test

- with the sorts of "word" input representations we see in RT corpora. e.g. punctuation and word still attached

```python
['If',
 'you',
 'were',
 'to',
 'journey',
 'to',
 'the',
 'North',
 'of',
 'England,',
 'you',
 'would',
 'come',
 'to',
 'a',
 'valley',
 'that',
 'is',
 'surrounded',
 'by',
 'moors',
 'as',
 'high',
 'as',
 'mountains.']
```

(I have to re-check how we most recently decided to handle surprisal of punctuation tokens in the aggregation; I'll get back to you on this.)

- use a model that attaches BOS/EOS tokens, e.g. `xlm-roberta`

```python
In [12]: t = transformers.AutoTokenizer.from_pretrained("xlm-roberta-base")

In [14]: t.convert_ids_to_tokens(t("this is a test").input_ids)
Out[14]: ['<s>', '▁this', '▁is', '▁a', '▁test', '</s>']
```

**Review by hans — CHANGES_REQUESTED** [review]
Sorry this is coming in delayed!

File: brainscore_language/models/huggingface.py (hunk 6 above)
Comment (mschrimpf) [reply brainscore_language/models/huggingface.py:193]: agree, I think this would be good to fix going forward

File: brainscore_language/models/huggingface.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mschrimpf) [reply brainscore_language/models/huggingface.py]: I'm a bit concerned about the model being able to output `nan` and then ignoring those predictions -- In an extreme case, the model could output all `nan`s except for a few words where it is very certain of the reading times and obtain a high score. My preference would be to output a non-nan value; we could do -1? That would signal non-regular output, but still allow it to be scored with a correlation metric without discarding values (that the model decides on).

[the remark this answers is not in the document]
Comment (hans) [reply brainscore_language/models/huggingface.py]: Copy-paste from Slack convo for posterity

> I see! That solves a problem but creates more, I think. The evaluation will be biased by any numeric value if we are using pearsonr or linear regression, such that the model results aren't interpretable (e.g. in a regression, the coefficient for surprisal outputs no longer really means surprisal).
As a benchmark dev I'd just pick through the model outputs and remove the -1 values manually so that my results are interpretable. Only difference then is that unaware benchmarkers might let this slip through without realizing.

> In any case models and benchmarks both will be doing data exclusions, right? Idk about the brain mapping, but in an RT benchmark for example, we usually drop 1) words at start and end of visual line, 2) words with punctuation. And models might reasonably want to drop OOV words (if that's a thing for them).
Is there a way we can track exclusions from each side as part of the benchmark record?

> In any eval I do I'm usually dropping thousands of tokens. Actually, in our latest RT paper (Wilcox 2020 CogSci) we dropped any token for which any model in our fleet did not provide valid surprisal estimates. That's probably the best option, but very hard to enforce on an evolving platform

> Suggested options summary
> 1. Explicitly track exclusions made by the model and by the benchmark; report this (e.g. as "coverage") alongside performance
> 2. Enforce that all models are evaluated on the same subset

**Review by hans — APPROVED** [review]
